### PR TITLE
Re-working Ticket Handling

### PR DESCRIPTION
**Pull request by hannestschofenig**
# Ticket Handling in TLS 1.3 

The TLS 1.3 specification harmonizes three pre-TLS 1.3 concepts: session resumption, session resumption without server-side state and PSK-based ciphers.
TLS 1.3 instead uses a design that introduces a new, post-handshake message (NewSessionTicket) and combines it with a PSK-based key exchange method. External PSK require a PSK identity and a secret to be configured while the NewSessionTicket message allows the configuration of the PSK identity as well as the secret dynamically following a successful handshake.

The NewSessionTicket message has the following structure: 

~~~
      struct {
          uint32 ticket_lifetime;
          uint32 ticket_age_add;
          opaque ticket_nonce<0..255>;
          opaque ticket<1..2^16-1>;
          Extension extensions<0..2^16-2>;
      } NewSessionTicket;
~~~

In addition to the ticket, which is opaque to the client, the server provides meta-data, which is necessary for proper operation when the client initiates a new handshake using this credential. RFC 8446 does not mandate a structure for the ticket, the Mbed TLS implementation defines one, which can be found in ssl_tls.c - see ssl_session_save(). The server stores the following content inside the ticket before encrypting it with a symmetric key: 

  - ticket creation time
  - ciphersuite
  - ticket lifetime
  - ticket-age-add parameter
  - key length
  - flags
  - resumption key
  - Peer's end-entity certificate
 
Note that the key used to encrypt the ticket by the server is known only to the server and it is rotated on a regular basis. 

The use of tickets is enabled with a compile-time switch (MBEDTLS_SSL_NEW_SESSION_TICKET) and via command line options (tickets) for use with the two example programs, ssl_client2 and ssl_server2. The mbedtls_ssl_ticket_init(), mbedtls_ssl_ticket_setup() and the mbedtls_ssl_conf_session_tickets_cb() API calls initialize data structures, configure crypto algorithms and register callbacks. Example callback functions for writing and parsing tickets are provided in ssl_ticket.h, ssl.h and ssl_ticket.c.

On the server-side the lifetime of tickets can be configured via the command line option ticket_timeout. On the client-side it is necessary to force the example program to run two handshakes via the reconnect options, of which there are several (reconnect, exchanges, reco_delay, reco_mode, and reconnect_hard).

For the server-side the NewSessionTicket message is sent immediately after the handshake. APIs to allow applications to trigger the transmission of NewSessionTicket messages will be added in the future. The ssl_write_new_session_ticket() call triggers the generation of the ticket. To pass the necessary parameters into the mbedtls_ssl_ticket_write() function, which internally calls the mbedtls_ssl_session_save()/ssl_session_save() functions (see ssl_tls.c) the session context contains data items to be included in the ticket itself plus the endpoint and minor_ver information. endpoint is needed to distinguish the client and the server functionality and minor_ver allows to place different information into the ticket depending on whether the handshake negotiated TLS 1.3 or an earlier TLS version. 

Once the client has received the ticket and re-uses it in a subsequent exchange, the server has to decrypt and process the information contained inside the ticket. The exact client-side processing will follow later. The ticket is presented to the server in the pre_shared_key extension as part of the ClientHello message. The pre_shared_key extension has the following structure:

~~~
      struct {
          opaque identity<1..2^16-1>;
          uint32 obfuscated_ticket_age;
      } PskIdentity;

      opaque PskBinderEntry<32..255>;

      struct {
          PskIdentity identities<7..2^16-1>;
          PskBinderEntry binders<33..2^16-1>;
      } OfferedPsks;

      struct {
          select (Handshake.msg_type) {
              case client_hello: OfferedPsks;
              case server_hello: uint16 selected_identity;
          };
      } PreSharedKeyExtension;
~~~

The mbedtls_ssl_parse_client_psk_identity_ext() API is responsbile for parsing the received structure. First, we check whether the identity matches the externally configured PSK value (which is stored in ssl->conf->p_psk). If this lookup was unsuccessful, then we attempt to treat the content of the PSK identity as the ticket and parse it with mbedtls_ssl_parse_psk_identity(). This function calls mbedtls_ssl_ticket_parse() and subsequently mbedtls_ssl_session_load()/ssl_session_load() and, after successful decryption, populates the values into the session structure. Then, various checks are made to determine whether the ticket is still valid (e.g. expired). 

The functionality on the client side is different because tickets are passed through clients as opaque blobs. Ticket support is enabled in ssl_client2 (in addition to the C pre-processor directives in config.h mentioned above) via mbedtls_ssl_conf_session_tickets(). A call to mbedtls_ssl_read() may return a MBEDTLS_ERR_SSL_RECEIVED_NEW_SESSION_TICKET error code, which offers the client an indication that a NewSessionTicket message has been received (instead of the expected application data). Depending on the reconnection mode selected via the command line parameter reco_mode two approaches are possible, namely to copy session, or to serialize a session. Serialization allows storing the ticket on some persistent storage in which case the ticket is first stored in a buffer of type unsigned char using mbedtls_ssl_session_save(). Without serialization, the mbedtls_ssl_get_session() API is used instead, which merely creates a copy of the session structure. The use of the pre-TLS 1.3 mbedtls_ssl_session_xxx APIs allows for better code reuse although more memory is needed using this approach because the TLS 1.3 ticket approach does not require a copy of the session structure. 

When a new handshake is trigger, which uses the previously distributed ticket, the stored information is recovered using mbedtls_ssl_session_load() and mbedtls_ssl_set_session(). mbedtls_ssl_session_load() restores the session state from the ticket buffer and mbedtls_ssl_set_session applies the session state into the ssl content. Finally, it is necessary to change the key exchange method to MBEDTLS_SSL_TLS13_KEY_EXCHANGE_MODE_PSK_DHE_KE or MBEDTLS_SSL_TLS13_KEY_EXCHANGE_MODE_PSK_KE so that the stack knows to use the PSK-based key exchange mode. 

When the handshake is then executed, the pre_shared_key extension will contain the ticket in the identity field, as described earlier. 

Note: This PR replaces https://github.com/hannestschofenig/mbedtls/pull/27

### CODE REVIEW

**Comment by hanno-becker** [comment]
@hannestschofenig What's the status of this work? Is it ready to be reviewed?

**Comment by hanno-becker** [comment]
@hannestschofenig There's a conflict in `ssl_tls13_generic.`. Could you address this?

File: include/mbedtls/ssl.h [hunk 1]
```diff
@@ -2978,6 +2858,9 @@ void mbedtls_ssl_conf_session_cache( mbedtls_ssl_config *conf,
         int (*f_set_cache)(void *, const mbedtls_ssl_session *) );
 #endif /* MBEDTLS_SSL_SRV_C */
 
+#if defined(MBEDTLS_SSL_PROTO_TLS1_2) || defined(MBEDTLS_SSL_PROTO_TLS1_1) || \
+    defined(MBEDTLS_SSL_PROTO_TLS1) || ( defined(MBEDTLS_SSL_NEW_SESSION_TICKET) && defined(MBEDTLS_SSL_SRV_C) )
+
```
Comment (lhuang04) [inline include/mbedtls/ssl.h:2863]: Do we need this macro?
`mbedtls_ssl_session_save` is used in `ssl_client2.c`. Does it mean that we can't use `mbedtls_ssl_session_save` even when `MBEDTLS_SSL_NEW_SESSION_TICKET` is defined?

File: include/mbedtls/ssl.h [hunk 2]
```diff
@@ -1114,31 +1091,31 @@ struct mbedtls_ssl_session
...
-    unsigned char* ticket;          /*!< TLS 1.3 session ticket acting as psk identity */
-    size_t ticket_len;              /*!< ticket length   */
-    uint32_t ticket_lifetime;       /*!< ticket lifetime hint    */
-    mbedtls_ssl_ticket_flags flags; /*!< ticket flags */
-    uint32_t ticket_age_add;        /* A randomly generated 32-bit value that is used to obscure the age of the ticket */
-    unsigned char* ticket_nonce;    /*!< ticket nonce value */
-    uint8_t ticket_nonce_len;       /*!< ticket nonce length */
-    size_t key_len;                 /*!< psk key length */
+    int minor_ver;                  /*!< Version information for selecting ticket format */
+    unsigned int endpoint : 1;      /*!< 0: client, 1: server               */
```
Comment (lhuang04) [inline include/mbedtls/ssl.h:1102]: Can we just use `ssl->conf->endpoint`? Is there any case that we may store ticket from both client and server?

File: library/ssl_tls.c [hunk 3]
```diff
@@ -5657,6 +5660,8 @@ static unsigned char ssl_serialized_session_header[] = {
     ( SSL_SERIALIZED_SESSION_CONFIG_BITFLAG >> 0 ) & 0xFF,
 };
 
+#if defined(MBEDTLS_SSL_SESSION_TICKETS) || ( defined(MBEDTLS_SSL_NEW_SESSION_TICKET) && defined(MBEDTLS_SSL_SRV_C) )
+
 /*
```
Comment (lhuang04) [inline library/ssl_tls.c:5665]: Similar to the comments in `ssl.h`, this macro seems mismatch with what is in the header.  And it also leaves `ssl_serialized_session_header` unused.

File: library/ssl_tls.c [hunk 4]
```diff
@@ -5690,6 +5698,17 @@ static unsigned char ssl_serialized_session_header[] = {
...
+ *
+ * In TLS/DTLS 1.3 the ticket contains the following content:
+ *
+ *  - ticket creation time ( start )
+ *  - ciphersuite ( ciphersuite )
+ *  - ticket lifetime ( ticket_lifetime )
+ *  - ticket-age-add parameter ( ticket_age_add )
+ *  - key length ( key_len )
+ *  - flags ( flags )
+ *  - key ( key )
```
Comment (lhuang04) [inline library/ssl_tls.c:5710]: Also add `ticket_received ` and `ticket` into the comments?

File: library/ssl_tls.c [hunk 5]
```diff
@@ -5933,49 +6088,161 @@ static int ssl_session_load( mbedtls_ssl_session *session,
         p += sizeof( ssl_serialized_session_header );
     }
 
-    /*
-     * Time
-     */
+#if defined(MBEDTLS_SSL_NEW_SESSION_TICKET) && defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL)
+
+    minor_ver = MBEDTLS_SSL_MINOR_VERSION_4; /* TBD: For testing only */
```
Comment (lhuang04) [inline library/ssl_tls.c:6093]: Do we have more plan on this? Sounds like we didn't use the `minor_ver` stored in the ticket.

File: library/ssl_tls13_client.c [hunk 6]
```diff
@@ -1015,22 +1021,43 @@ int mbedtls_ssl_write_pre_shared_key_ext( mbedtls_ssl_context *ssl,
...
-        {
-            MBEDTLS_SSL_DEBUG_MSG( 1, ( "mbedtls_hash_size_for_ciphersuite == -1, mbedtls_ssl_write_pre_shared_key_ext failed" ) );
-            return( MBEDTLS_ERR_SSL_INTERNAL_ERROR );
+            if( hash_len == 0 )
+            {
+                MBEDTLS_SSL_DEBUG_MSG( 1, ( "mbedtls_hash_size_for_ciphersuite == 0, mbedtls_ssl_write_pre_shared_key_ext failed" ) );
+                return( MBEDTLS_ERR_SSL_INTERNAL_ERROR );
+            }
+            ssl->handshake->ciphersuite_info = suite_info;
         }
```
Comment (lhuang04) [inline library/ssl_tls13_client.c:1042]: This block (`resume == 1`) could be moved out of the `for` loop as it is not use anything from `ciphersuites`.

File: library/ssl_tls13_client.c [hunk 7]
```diff
@@ -1087,33 +1124,33 @@ int mbedtls_ssl_write_pre_shared_key_ext( mbedtls_ssl_context *ssl,
...
         /* Calculate obfuscated_ticket_age */
         /* ( but not for externally configured PSKs ) */
-        if( ssl->conf->ticket_age_add > 0 )
+        if( ssl->session_negotiate->ticket_age_add > 0 )
         {
 #if defined(MBEDTLS_HAVE_TIME)
             now = time( NULL );
 
-            if( !( ssl->conf->ticket_received <= now && now - ssl->conf->ticket_received < 7 * 86400 * 1000 ) )
+            if( !( ssl->session_negotiate->ticket_received <= now && now - ssl->session_negotiate->ticket_received < 7 * 86400 * 1000 ) )
```
Comment (lhuang04) [inline library/ssl_tls13_client.c:1146]: This is not related to this PR. It sounds like the `now` is in second, and the `7 * 86400 * 1000` should be in millisecond, otherwise this means the ticket won't be expired after 1000 weeks.

**Comment by hanno-becker** [comment]
@hannestschofenig I'd like to understand the top-level API and its usage better before diving into the precise code.

* External PSKs: If I understand correctly, we aim to reuse the existing PSK configuration API for external PSKs. This has two minor problems: (a) We can only configure one external PSK for now. (b) A PSK is always associated with a hash function, and the current PSK configuration API doesn't allow us to specify that. Following the standard, we could explain that external PSKs are currently always bound to SHA-256.
* Resumption PSKs: As mentioned in the [mailing list post](https://lists.trustedfirmware.org/pipermail/mbed-tls/2020-April/000049.html), we need to explain the new semantics of `mbedtls_ssl_session_load/save()` if we want to re-use it for session tickets. The pre-1.3 semantics is idempotent - it doesn't matter how often you call those functions, they'll always to the same - but with the advent of multiple session tickets (both in terms of what the client can suggest and in terms of what the server can send), it needs to be considered whether we want to change this semantics.

**Comment by hannestschofenig** [comment]
Replaced by https://github.com/hannestschofenig/mbedtls/pull/152
